### PR TITLE
test: check code with ESLint

### PR DESCRIPTION
**Pull request by EvanHahn**
`npm run lint:eslint` (run as part of `npm lint` and `npm test`) now checks our code with ESLint.

This also fixes or disables all lint failures.

### CODE REVIEW

File: eslint.config.js [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+  ...tseslint.configs.recommended,
+  {
+    ignores: ['dist/', 'generated/', 'intermediate/'],
+  },
+  {
+    rules: {
+      '@typescript-eslint/ban-ts-comment': [
+        'error',
+        {
+          'ts-expect-error': false,
```
Comment (EvanHahn) [inline eslint.config.js:18]: The default still allows this, but makes us leave a comment.

File: eslint.config.js [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+        },
+      ],
+      '@typescript-eslint/no-explicit-any': 'off',
```
Comment (EvanHahn) [inline eslint.config.js:21]: We use a lot of `any`s so I don't think we should disable this right now.

File: eslint.config.js [hunk 3]
```diff
@@ -0,0 +1,31 @@
...
+      '@typescript-eslint/no-unused-vars': [
+        'error',
+        {
+          varsIgnorePattern: '^_',
+          argsIgnorePattern: '^_',
+        },
+      ],
+    },
```
Comment (EvanHahn) [inline eslint.config.js:29]: Matches rules we have elsewhere.

File: src/lib/utils.ts [hunk 4]
```diff
@@ -74,18 +74,16 @@ export function parseVersionId(versionId: string): VersionIdObject {
...
-    updatedAt,
-    deleted,
+    docId: _docId,
+    versionId: _versionId,
+    originalVersionId: _originalVersionId,
+    links: _links,
+    forks: _forks,
+    createdAt: _createdAt,
+    updatedAt: _updatedAt,
+    deleted: _deleted,
```
Comment (gmaclennan) [inline src/lib/utils.ts:85]: I'm not a huge fan of re-writing code to satisfy a linter, although this may just be a personal preference.

When reviewing or writing code like this, it requires contextual knowledge of the eslint configuration to know why the code is written this way. Similarly a future developer might write variables prepended with a `_` and not be expecting eslint rules to be ignored because of it.

My preference is to use eslint ignore comments because it is more explicit about what the developer was intending.

File: test/fixtures/good-docs-completed.js [hunk 5]
```diff
@@ -205,8 +205,7 @@ export const goodDocsCompleted = [
       updatedAt: cachedValues.updatedAt,
       links: [],
       name: 'my device name',
-      // @ts-expect-error
-      deviceType: 'motorbike',
+      deviceType: /** @type {any} */ ('motorbike'),
```
Comment (gmaclennan) [inline test/fixtures/good-docs-completed.js:208]: Why change this? In the (extremely unlikely) scenario that we made `motorbike` a valid device type, we would want a @ts-expect-error to show an error here, because the test depends on this being a TS error.

File: src/lib/utils.ts (hunk 4 above)
Comment (EvanHahn) [reply src/lib/utils.ts]: Updated in eabe017905823df5bb5a5cd11cbbfa581d6a5cc3.

File: test/fixtures/good-docs-completed.js (hunk 5 above)
Comment (EvanHahn) [reply test/fixtures/good-docs-completed.js]: Removed in b1e28d4f9f9fd3a7ac176f6dcb5d04edc5191b2c.

**Review by gmaclennan — APPROVED** [review]
Looks good. Small niggle, can be ignored: we use [`filter-obj`](https://github.com/sindresorhus/filter-obj) in a couple of other modules, and I would lean towards using a well-maintained and tested module like that rather than our own implementation, which is doesn't have tests for edge-cases, e.g. symbol properties.

**Comment by EvanHahn** [comment]
I'll plan do that in a followup.
